cmd/restic/backup_progress.go

```diff
@@ -0,0 +1,176 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"text/template"
+	"time"
+)
+
+// ProgressAction type of progress output
+type ProgressAction string
+
+const (
+	// ProgressTypeScanner scanner progress type
+	ProgressTypeScanner ProgressAction = "scanner"
+	// ProgressTypeArchive archive process progress type
+	ProgressTypeArchive ProgressAction = "archive"
+)
+
+type (
+	// ProgressStatus provides json archive process progress
+	ProgressStatus struct {
+		ProgressType   ProgressAction `json:"type"`
+		Duration       time.Duration  `json:"duration"`
+		Directories    uint64         `json:"directories"`
+		Files          uint64         `json:"files"`
+		Completed      uint64         `json:"completed"`
+		Total          uint64         `json:"total"`
+		Bps            uint64         `json:"bps"`
+		CompletedBytes uint64         `json:"completed_bytes"`
+		ItemsDone      uint64         `json:"items_done"`
+		ItemsTotal     uint64         `json:"items_total"`
+		TotalBytes     uint64         `json:"total_bytes"`
+		ETA            uint64         `json:"eta"`
+		Errors         uint64         `json:"errors"`
+		Seconds        uint64         `json:"seconds"`
+	}
+)
+
+// Templates for outputting current status
+const (
+	ScanProgressUpdateTemplate      = `[{{duration .Duration}}] {{.Directories}} directories, {{.Files}} files, {{bytes .TotalBytes}}`
+	ScanProgressDoneTemplate        = `scanned {{.Directories}} directories, {{.Files}} files in {{duration .Duration}}\n`
+	ArchiveProgressUpdateTemplate   = `[{{duration .Duration}}] {{percent .Completed .Total}}  {{bytes .Bps}}/s  {{bytes .CompletedBytes}} / {{bytes .TotalBytes}}  {{.ItemsDone}} / {{.ItemsTotal}} items  {{.Errors}} errors  `
+	ArchiveProgressEstimateTemplate = `ETA {{seconds .ETA}}`
+	ArchiveProgressDoneTemplate     = `\nduration {{duration .Duration}}, {{rate .TotalBytes .Duration}}\n`
+)
+
+var scanProgressUpdateTemplate,
+	scanProgressDoneTemplate,
+	archiveProgressTemplate,
+	archiveProgresEstimateTemplate,
+	archiveProgressDoneTemplate *template.Template
+
+func init() {
+	templateFunctions := template.FuncMap{
+		"duration": formatDuration,
+		"percent":  formatPercent,
+		"bytes":    formatBytes,
+		"seconds":  formatSeconds,
+		"rate":     formatRate,
+	}
+
+	scanProgressUpdateTemplate, _ = template.New("scanProgressUpdateTemplate").Funcs(templateFunctions).Parse(ScanProgressUpdateTemplate)
```

Comment:
This throws away the error, use `template.Must()` around it (which will panic if an error occurs).